src/lib/define-dynamic-block.js

```diff
@@ -176,14 +219,64 @@ const defineDynamicBlock = (ScratchBlocks, categoryInfo, staticBlockInfo, extend
                 }
                 return `%${++argCount}`;
             });
+            const wasRendered = this.rendered;
+            this.rendered = false;
+            const connectionMap = this.disconnectOldBlocks_();
+            this.removeAllInputs_();
+
             this.interpolate_(scratchBlocksStyleText, args);
+
+            // Interpolate takes care of laying out the base block,
+            // now we need to reconnect blocks into the inputs
+            // with the same name.
+            let inputIndex = -1;
+            args.forEach(arg => {
+                let type;
+                if (arg.type === 'input_value') {
+                    inputIndex++;
+                    if (arg.check && arg.check === 'boolean') {
+                        type = 'b';
+                    } else {
+                        type = 's';
+                    }
+                    if (inputIndex < this.inputList.length) {
+                        const input = this.inputList[inputIndex];
+                        this.populateArgument_(type, inputIndex, connectionMap, input.name, input);
+                    }
+                }
+            });
+
+            // Set values on any args that have selectedValue specified
+            args.forEach(arg => {
+                if (arg.type === 'field_dropdown' && blockInfo.arguments[arg.name].selectedValue) {
+                    const field = this.getField(arg.name);
+                    if (field) {
+                        field.setValue(blockInfo.arguments[arg.name].selectedValue);
+                    }
+
+                }
+            });
+
+            this.rendered = wasRendered;
+            if (wasRendered && !this.isInsertionMarker()) {
+                this.initSvg();
+                this.render();
+            }
         },
         setBlockInfo: function (blockInfo) {
             this.needsBlockInfoUpdate = true;
             this.blockInfoText = JSON.stringify(blockInfo);
             const mutation = this.mutationToDom();
             this.domToMutation(mutation);
-        }
+        },
+        // Extra functions for dynamically updating the layout of a block and reconnecting
+        // inputs and shadow blocks.
+        removeAllInputs_: ScratchBlocks.ScratchBlocks.ProcedureUtils.removeAllInputs_,
```

Comment:
It's not great that this and the following functions are being borrowed from ScratchBlocks.ProcedureUtils, but maybe we can move those scratch-blocks functions into a more generic util post-extensionification.

Reply:
I agree that it looks a bit... not-great... but if all these methods expect to be called with the block as `this` then your current approach might be the best option. If that's the case then I'm inclined to just call this a consequence of the scratch-blocks API.

Reply:
> If that's the case then I'm inclined to just call this a consequence of the scratch-blocks API.

I think I'm resigned to this statement. These are the woes of dynamically updating a block...